src/interfaces/chain.cpp

```diff
@@ -148,22 +148,12 @@ class LockImpl : public Chain::Lock, public UniqueLock<RecursiveMutex>
     using UniqueLock::UniqueLock;
 };
 
-class NotificationsHandlerImpl : public Handler, CValidationInterface
+class NotificationsProxy : public CValidationInterface
 {
 public:
-    explicit NotificationsHandlerImpl(Chain& chain, Chain::Notifications& notifications)
-        : m_chain(chain), m_notifications(&notifications)
-    {
-        RegisterValidationInterface(this);
-    }
-    ~NotificationsHandlerImpl() override { disconnect(); }
-    void disconnect() override
-    {
-        if (m_notifications) {
-            m_notifications = nullptr;
-            UnregisterValidationInterface(this);
-        }
-    }
+    explicit NotificationsProxy(std::shared_ptr<Chain::Notifications> notifications)
+        : m_notifications(std::move(notifications)) {}
+    virtual ~NotificationsProxy() = default;
```

Comment:
Is the `NotificationsProxy` class intended for subclassing? If not, why its destructor is declared with `virtual`?

Reply:
Removed.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/18338#discussion_r399674652

> nit: Is the `NotificationsProxy` class intended for subclassing? If not, why its destructor is declared with `virtual`?

Would again suggest reverting 17316b3a6eb41204824111ea3cfeb2e1b7d80b1b to 41b0baf43c243b64b394e774e336475a489cca2b to undo this change. The point of declaring this virtual is to prevent memory leaks and incomplete destruction in case this is subclassed (whether intended or not). From https://isocpp.org/wiki/faq/virtual-functions#virtual-dtors:

> Here’s a simplified rule of thumb that usually protects you and usually doesn’t cost you anything: make your destructor virtual if your class has any virtual functions. 

This is an important rule to follow here and other places. Other alternatives would be to declare the class final or give it a protected nonvirtual destructor, but these are optimizations that are more fragile. Just having a virtual destructor when there are virtual functions is usually the simplest safe approach.

Reply:
Sorry for noise. I forgot that member functions defined with `override` are still `virtual`.

Reply:
@ryanofsky ibid:
> Note: in a derived class, if your base class has a virtual destructor, your own destructor is automatically virtual.

Reply:
> @ryanofsky ibid:
> 
> > Note: in a derived class, if your base class has a virtual destructor, your own destructor is automatically virtual.

Yes, the problems happen when code changes over time. Without the override keyword somebody could make a change to the base class removing `virtual` and unknowingly cause memory leaks in the derived class. But with the override keyword, the compiler will show a "does not override" error building the derived class and not allow this.

Reply:
Yes, I [understand](16710) that consistent usage of `override` in derived classes protects from bugs caused by changes in base classes over time.

But how `virtual` could help for a desctructor in a derived class, if:
> in a derived class, if your base class has a `virtual` destructor, your own destructor is _automatically_ `virtual`. You might need an explicitly defined destructor for other reasons, but there’s _no need_ to redeclare a destructor simply to make sure it is `virtual`. No matter whether you declare it with the `virtual` keyword, declare it without the `virtual` keyword, or don’t declare it at all, it’s _still_ `virtual`.

?

Reply:
> But how `virtual` could help for a desctructor in a derived class, if:
> 
> > in a derived class, if your base class has a `virtual` destructor, your own destructor is _automatically_ `virtual`. You might need an explicitly defined destructor for other reasons, but there’s _no need_ to redeclare a destructor simply to make sure it is `virtual`. No matter whether you declare it with the `virtual` keyword, declare it without the `virtual` keyword, or don’t declare it at all, it’s _still_ `virtual`.
> 
> ?

None of that is applicable here. NotificationsProxy base class is CValidationInterface which does not have a virtual destructor. If it did have a virtual destructor, this could be declared with `override` instead of `virtual`, and registration would no longer require shared pointers with custom deleters. But these changes would increase size of the bugfix and are better saved for a followup PR. This PR is just trying to do the simple thing and

- Declare virtual destructors where there are virtual functions
- Use `override` instead of `virtual` where it prevents derived class code from being silently broken by future changes to the base class

Reply:
Oh, I missed that: https://github.com/bitcoin/bitcoin/blob/41b0baf43c243b64b394e774e336475a489cca2b/src/validationinterface.h#L77-L83